dspy/evaluate/evaluate.py

```diff
@@ -93,27 +135,110 @@ def _execute_multi_thread(self, wrapped_program, devset, num_threads, display_pr
-        pbar.set_description(f"Average Metric: {ncorrect} / {ntotal}  ({round(100 * ncorrect / ntotal, 1)})")
+    @staticmethod
+    def _update_progress(pbar, ncorrect, ntotal):
+        pbar.set_description(
+            f"Average Metric: {ncorrect} / {ntotal}  ({round(100 * ncorrect / ntotal, 1)})"
+        )
         pbar.update()
 
+    ### Type stubs below for three alternative calling patterns, depending
+    ### on the values of `return_all_scores` and `return_outputs`
+    @overload
+    def __call__(
+            self,
+            program,
+            metric=...,
+            devset: Optional[Iterable] = ...,  # Needs more specific type, if possible
+            num_threads: Optional[int] = ...,
+            display_progress: Optional[bool] = ...,
+            display_table: Union[bool, int, None] = ...,
+            *,
+            return_all_scores: Literal[True],
+            return_outputs: Literal[True],
+    ) -> Tuple[float, List, List[float]]:
+        ...
+
+    @overload
+    def __call__(
+            self,
+            program,
+            metric=...,
+            devset: Optional[Iterable] = ...,  # Needs more specific type, if possible
+            num_threads: Optional[int] = ...,
+            display_progress: Optional[bool] = ...,
+            display_table: Union[bool, int, None] = ...,
+            *,
+            return_all_scores: Literal[True],
+            return_outputs: Literal[False],
+    ) -> Tuple[float, List[float]]:
+        ...
+
+    @overload
+    def __call__(
+            self,
+            program,
+            metric=...,
+            devset: Optional[Iterable] = ...,  # Needs more specific type, if possible
+            num_threads: Optional[int] = ...,
+            display_progress: Optional[bool] = ...,
+            display_table: Union[bool, int, None] = ...,
+            *,
+            return_all_scores: Literal[False],
+            return_outputs: Literal[True],
+    ) -> Tuple[float, List[Tuple[Example, Prediction, Any]]]:
+        ...
+
+    @overload
+    def __call__(
+            self,
+            program,
+            metric=...,
+            devset: Optional[Iterable] = ...,  # Needs more specific type, if possible
+            num_threads: Optional[int] = ...,
+            display_progress: Optional[bool] = ...,
+            display_table: Union[bool, int, None] = ...,
+            *,
+            return_all_scores: Literal[False],
+            return_outputs: Literal[False],
+    ) -> float:
+        ...
+
     def __call__(
-        self,
-        program,
-        metric=None,
-        devset=None,
-        num_threads=None,
-        display_progress=None,
-        display_table=None,
-        return_all_scores=False,
-        return_outputs=False,
+            self,
+            program,
+            metric: Optional[Callable] = None,
+            devset: Optional[Iterable] = None,  # Needs more specific type, if possible
```

Comment:
remove comment

Reply:
Should I refine these type hints?  I believe they should be:

``` python
metric: Optional[Callable[[Iterable[Example], Iterable[Prediction], MetricValue] = None
devset: Optional[Iterable[Example]] = None
```
with `MetricValue` defined as `Union[bool, float]`.

Reply:
Actually, I'm a little confused about what's going on here with respect to `return_scores`.  I have said that `__call__` returns a list of floats for `return_scores`, but metric functions can return boolean values, so maybe the right type for this return value is `Union[List[bool], List[float]]`

I'm not sure if booleans are coerced to floats at some point

Reply:
I think the latter`Union[List[bool], List[float]]` makes sense for this!